File: 1269. 停在原地的方案数/动态规划.swift
class Solution {
    func numWays(_ steps: Int, _ arrLen: Int) -> Int {
        let arrLen = min(arrLen, steps + 1)
        // table[step: 0...steps][idx: 0..<arrLen]
        var table = [[Int]](repeating: [Int](repeating: -1, count: arrLen), count: steps + 1)

        func memory(_ step: Int, _ idx: Int) -> Int {
            if idx < 0 || idx >= arrLen {
                return 0
            }
            if table[step][idx] == -1 {
                table[step][idx] = search(step, idx) % 1000000007
            }
            return table[step][idx]
        }

        func search(_ step: Int, _ idx: Int) -> Int {
            if step == steps {
                return idx == 0 ? 1 : 0
            }
            let step = step + 1
            return memory(step, idx - 1) + memory(step, idx) + memory(step, idx + 1)
        }

        return memory(0, 0)
    }
}

File: 0664. 奇怪的打印机/动态规划.js
// 244ms
var strangePrinter = function(s) {
    const len = s.length
    let table = new Array(s.length).fill(0).map(_ => new Array(s.length).fill(-1))

    function memory(l, r) {
        if (table[l][r] == -1) {
            table[l][r] = search(l, r)
        }
        return table[l][r]
    }

    function search(l, r) {
        if (l == r) {
            return 1
        }
        if (s[l] == s[r]) {
            return memory(l, r - 1)
        } else {
            let minn = Number.MAX_VALUE
            for (let m = l; m + 1 <= r; m += 1) {
                minn = Math.min(minn, memory(l, m) + memory(m + 1, r))
            }
            return minn
        }
    }

    return memory(0, len - 1)
};